js/src/Zone.js

```diff
@@ -0,0 +1,91 @@
+import ExpectedMark from "./ExpectedMark";
+import UXCapture from "./UXCapture";
+
+/**
+ * Zone represents collection of elements groupped together and corresponding phase of page load
+ *
+ * Only one set of zones can be tracked for the same view at one point in time
+ */
+export default class Zone {
+  // constructs individual zone object
+  constructor(config) {
+    // {
+    //   name: "ux-destination-verified",
+    //   marks: ["ux-image-online-logo", "ux-image-inline-logo"]
+    //   selectors: ["img.logo"],
+    //   startMark: "navigationStart",
+    //   onMeasure: measureName => {}
+    //   onMark: markName => {}
+    // }
+
+    // name to be used for UserTiming measures
+    this.measureName = config.name;
+
+    // callback to execute when Zone is complete
+    this.onMeasure = config.onMeasure;
+
+    // callback for marks to call when they are complete
+    this.onMark = config.onMark;
+
+    // array of mark names that were recorded so far
+    this.recordedMarkNames = [];
+
+    this.startMarkName = config.startMarkName || "navigationStart";
+
+    // look up existing mark object or create a new one
+
+    // now just string names for UserTiming marks
+    // in the future, different types of events, e.g. ImageElement("logo"), PaintTimer("first-paint")
+    this.marks = config.marks.map(markName => {
+      const mark = ExpectedMark.create(markName);
+
+      mark.onComplete(mark => {
```

Comment:
React encourages a pattern that I think would be helpful here - instead of imperatively adding listeners to the `mark` with a static method, the listener can instead be passed in declaratively as an `onMark` 'prop' that gets assigned/merged into the mark's `onMarkListeners`. It wouldn't affect the top level API, but it would simplify the `ExpectedMark` interface, and would parallel the `onMark`/`onMeasure` callback passing interface that `Zone` has, which is good for consistency/understandability, particularly for React developers.

The main interface change is that `name` and `onMark` would both have to be passed into the `ExpectedMark` `create` method and constructor

```js
// Zone constructor
// compose `this.onMark` from `config.onMark` _plus additional 'complete' logic_
this.onMark = (markName => {
  config.onMark(markName);
  this.checkCompletion(markName) && this.complete(markName);
}).bind(this);

// `this.marks` assignment becomes a simpler mapping, with each `ExpectedMark` getting a `this.onMark` callback
this.marks = config.marks.map(name => ExpectedMark.create({ name, onMark: this.onMark }));
```

`ExpectedMark` would update to manage `this.onMarkListeners` directly in the constructor and `create` static methods

```js
static create(config) {
  let mark = ExpectedMark.get(config.name);
  if (mark) {
    mark.onMarkListeners.push(config.onMark || NOOP);
  } else {
    mark = new ExpectedMark(config);
    _expectedMarks.push(mark);
  }
   return mark;
}

constructor(config) {
  this.name = config.name;
  this.onMarkListeners = [config.onMark || NOOP];
}
```

Reply:
Let's do this in a followup. I am worried also that we have external API with in `UX.config({onMark, onMeasure})` that is different from internal `onMark` callback on `ExpectedMark` object and it might be confusing the use-case here.

Reply:
I'll create a ticket for this.